kernel/boot_info/src/multiboot2.rs

```diff
@@ -0,0 +1,176 @@
+use crate::ElfSectionFlags;
+use core::{cmp, iter::Iterator, ops::Range};
+use kernel_config::memory::KERNEL_OFFSET;
+use memory_structs::{PhysicalAddress, VirtualAddress};
+
+impl<'a> crate::MemoryRegion for &'a multiboot2::MemoryArea {
+    fn start(&self) -> PhysicalAddress {
+        PhysicalAddress::new_canonical(self.start_address() as usize)
+    }
+
+    fn len(&self) -> usize {
+        multiboot2::MemoryArea::size(self) as usize
+    }
+
+    fn is_usable(&self) -> bool {
+        matches!(self.typ(), multiboot2::MemoryAreaType::Available)
+    }
+}
+
+type MemoryRegionIterator<'a> = impl Iterator<Item = &'a multiboot2::MemoryArea>;
+
+pub struct MemoryRegions<'a> {
+    inner: MemoryRegionIterator<'a>,
+}
+
+impl<'a> Iterator for MemoryRegions<'a> {
+    type Item = &'a multiboot2::MemoryArea;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.next()
+    }
+}
+
+impl crate::ElfSection for multiboot2::ElfSection {
+    fn name(&self) -> &str {
+        multiboot2::ElfSection::name(self)
+    }
+
+    fn start(&self) -> VirtualAddress {
+        VirtualAddress::new_canonical(self.start_address() as usize)
+    }
+
+    fn len(&self) -> usize {
+        multiboot2::ElfSection::size(self) as usize
+    }
+
+    fn flags(&self) -> ElfSectionFlags {
+        ElfSectionFlags::from_bits_truncate(multiboot2::ElfSection::flags(self).bits())
```

Comment:
this implicitly relies on the `ElfSectionFlags` bit layout to be identical to that of `multiboot2::ElfSection`. 

I think this is okay, but that requirement is sneakily hidden on only expressed here if you look closely.
For ex, I didn't know that until I studied this line.

One way to encode this explicitly would be to directly use the `multiboot2::ElfSection` const flag definition's bit values in the definition of `ElfSectionFlags`, i.e.,:
```rust
    pub struct ElfSectionFlags: u64 {
        /// The section contains data that should be writable during program execution.
        const WRITABLE = multiboot2::ElfSectionFlags::WRITABLE.bits();

        /// The section occupies memory during the process execution.
        const ALLOCATED = multiboot2::ElfSectionFlags::ALLOCATED.bits();

        /// The section contains executable machine instructions.
        const EXECUTABLE = multiboot2::ElfSectionFlags::EXECUTABLE.bits();
    }
```

That way, an implicitly incorrect conversion could never happen. Alternatively, you could code this using a more readable series of three conditionals, one per flag. That might mesh better with the upcoming UEFI definition of `ElfSectionFlags`, which I assume likely differs from multiboot2.

Reply:
The ELF section flags are the same regardless of the boot specification used. [Table 7--8](https://docs.oracle.com/cd/E19120-01/open.solaris/819-0690/6n33n7fcj/index.html). I do agree that using three conditionals would be more readable.

Reply:
> The ELF section flags are the same regardless of the boot specification used.

Oh true. For some reason I was thinking of bootloader-defined flag sets.

> I do agree that using three conditionals would be more readable, so I can do that if you think we should.

Sure, go for it.